rust/lance-index/src/scalar/ngram.rs

```diff
@@ -434,127 +577,434 @@ impl NGramIndexBuilder {
         Ok(())
     }
 
-    fn process_batch(&mut self, batch: &RecordBatch) {
+    async fn process_batch(&mut self, tokens_and_ids: Vec<(u32, u64)>) -> Result<()> {
+        let mut tokens_seen = 0;
+        for (token, row_id) in tokens_and_ids {
+            tokens_seen += 1;
+            // This would be a bit simpler with entry API but, at scale, the vast majority
+            // of cases will be a hit and we want to avoid cloning the string if we can.  So
+            // for now we do the double-hash.  We can simplify in the future with raw_entry
+            // when it stabilizes.
+            self.state
+                .tokens_map
+                .entry(token)
+                .or_default()
+                .insert(row_id);
+        }
+        self.tokens_seen += tokens_seen;
+        if self.tokens_seen >= self.options.tokens_per_spill {
+            let state = self.state.take();
+            self.flush(state).await?;
+        }
+        Ok(())
+    }
+
+    fn spill_filename(id: usize) -> String {
+        format!("spill-{}.lance", id)
+    }
+
+    fn tmp_spill_filename(id: usize) -> String {
+        format!("spill-{}.lance.tmp", id)
+    }
```

Comment:
do we need to worry about these filenames colliding with files generated by another process? For instance if someone tries to create two trigram indexes on a table at the same time.

Reply:
Good question but we should be ok.  Each call to train will create a new `NGramIndexBuilder`.  Each new `NGramIndexBuilder` gets its own spill store and uses a unique temp directory (returned from `tempfile::tempdir`)